packages/langium/src/grammar/validation/validator.ts

```diff
@@ -785,6 +790,160 @@ export class LangiumGrammarValidator {
         }
     }
 
+    /** This validation is specific for assignments with '=' as assignment operator and checks,
+     * whether the operator should be '+=' instead. */
+    checkAssignmentOperator(assignment: ast.Assignment, accept: ValidationAcceptor): void {
```

Comment:
I think you should check instead the parser rule and always traverse down.
With this approach it seems that you are always moving up to the parser rule and then go down to get the entire context. That could result in redundant work if there is no cache in-between.
I also think for the sake of maintainability, would this not be easier to go always downwards within the checker, starting at the parser rule root?

Reply:
In my imagination you just need a map from property names to a counter per rule alternative. Fragments need to be resolved of course, but even with this, I think, the code becomes a lot simpler.

Reply:
Yes, mixing upwards and downwards feels a bit strange for me as well.
(The reason for that is historic, I started in that way and the challenge grew and grew 🙂)

I like your idea to start at the parser rules and only going downwards together with a map for the found feature names. That would make it cleaner.

However, I am not sure about these details:
- Actions interrupt the straight-forward recursive search: Is it enough to handle it as "a new parser rule"? I assume yes ...
- What about fragments which are called in other Langium files? Is that no special case anymore or does it require special treatments, which is higher than the simplification before?

I will continue to think about this ... thanks for the suggestion!

Reply:
> map from property names to a counter per rule alternative

We need to store the `Assignment` nodes as well, otherwise, we cannot annotate the warning at the correct position.

What do you mean with "per rule alternative"?

Reply:
For a parser rule P ::= A|B|C you have 3 alternatives A, B and C.

```ts
//foreach alternative do
const map = new Map<string, Assignment[]>();
foreach(const assignment in alternative…) {
  if(!map.has(assignemnt.name)) { map.set(assignment.name, []) }
  map.get(assignment.name)!.push(assignment)
}
//check if any entry of map has more then 1 assignment
```

Reply:
I just pushed a big refactoring, most of the special cases vanished and the code should be more readable.

WDYT?